Launching a Workflow using an AnyWorkflow

morganzellers opened the pull request:
### Linked Issue: 

<!-- (See our contributing guidelines for more details) -->

<!-- Please remove any items below that may not apply to your Pull Request -->
## Checklist:
- [x] Did you sign the [Contributor License Agreement](https://cla-assistant.io/wwt/SwiftCurrent)?
- [x] Is the linter reporting 0 errors?
- [ ] Did you comply with our [styleguide](https://github.com/wwt/SwiftCurrent/blob/main/.github/STYLEGUIDE.md)?
- [ ] Is there [adequate test coverage](https://github.com/wwt/SwiftCurrent/blob/main/.github/CONTRIBUTING.md#test-etiquette) for your new code?
- [ ] Does the CI pipeline pass?
- [ ] Did you [update the documentation](https://github.com/wwt/SwiftCurrent/blob/main/.github/CONTRIBUTING.md#documentation)?
- [ ] Did you [update the sample app](https://github.com/wwt/SwiftCurrent/blob/main/.github/CONTRIBUTING.md#sample-app)?
- [ ] Do we need to [increment the minor/major version](https://github.com/wwt/SwiftCurrent/blob/main/.github/CONTRIBUTING.md#versioning)?
- [ ] Did you [change the public API](https://github.com/wwt/SwiftCurrent/blob/main/.github/CONTRIBUTING.md#public-api)?
- [ ] Have you done everything you can to make sure that errors that can occur are compile-time errors, and if they have to be runtime do you have adequate tests and documentation around those runtime errors? [For more details](https://github.com/wwt/SwiftCurrent/blob/main/.github/CONTRIBUTING.md#errors).

----

### If Applicable:
- [ ] Did you test when the first item is skipped?
- [ ] Did you test when the last item is skipped?
- [ ] Did you test when middle items are skipped?
- [ ] Did you test when incorrect data is passed forward?
- [ ] Did you test proceeding backwards?

----

### If Public API Has Changed:
- [ ] Did you deprecate (rather than remove) any old methods/variables/etc? [Our philosophy for deprecation](https://github.com/wwt/SwiftCurrent/blob/main/.github/CONTRIBUTING.md#deprecation).
- [ ] Have you done the best that you can to make sure that the compiler guides people to changing to the new API? (Example: the renamed attribute)
- [ ] If necessary, have you tested the upgrade path for at least N-1 versions? For example, if data persists between v1 and v2 then that upgrade should be tested and as easy as we can make it.

Richard-Gist reviewed (commented):
I'm going to try to stay hands-off, and let you run with this. Reach out if you want some help.

Tyler-Keith-Thompson commented on Tests/SwiftCurrent_SwiftUITests/SwiftCurrent_SwiftUITests.swift:
```diff
@@ -731,6 +731,30 @@ final class SwiftCurrent_SwiftUIConsumerTests: XCTestCase, App {
         wait(for: [exp], timeout: TestConstant.timeout)
     }
+
+    func testLaunchingAWorkflowFromAnAnyWorkflow() {
```
That's exactly what the generic constraint tests are for

morganzellers replied:
@Tyler-Keith-Thompson I'll take a look at those in the AM 👍🏻

morganzellers commented:
Added a test for fatalError when no items in a workflow earlier then spent most of the day deciphering some of the unit tests and working through adding needed `thenProceed`'s.

morganzellers commented:
EOD: Was able to drive out a few more `thenProceed`'s today. Starting to understand them more, but am still fuzzy at times on which situation in a workflow a thenProceed signature translates to. 

Also fuzzy on the difference between setting `typealias WorkflowOutput = String` vs `typealias WorkflowOutput = AnyWorkflow.PassedArgs`. I thought that setting the latter basically said "Whatever type I'm passed is what I'm passing on" but there were a few times today where I doubted my memory.

morganzellers commented on Sources/SwiftCurrent_SwiftUI/Extensions/WorkflowExtensions.swift:
```diff
@@ -36,14 +36,40 @@ extension Workflow where F: FlowRepresentable & View {
+//            return flowPersistence(cast)
+//        })
+//        return workflow
+//    }
```
Hate to commit this much commented code, but I didn't trust myself to remember the signatures I want to drive out next over the weekend 😬.

morganzellers commented:
EOD: Added support for startingArgs - swapped to using the GenericConstraintsTests (copied and renamed for ExtendedFlowRepresentableMetadata) to drive out more functionality for persistence and launch style arguements.

Richard-Gist commented:
I tried to work on this a bit today, but there's a LOT of work we'll need to do to convert all of the tests over to actually test things.  I tried some different routes like maybe taking the generic constraint tests from SwiftCurrent core and seeing if those were easier to change over, but they weren't because of a difference in what you can use as starting arguments.

Maybe someone with a bit more regex prowess can make mass edits to convert the code, but for now, the only way forward I see is essentially translating all of the tests.  I found I had to update how the workflow was created, how it was launched, and how it was asserted.

I did contemplate for a while if we even needed to do all of this, which is still a "maybe" to a somewhat of a "yeah".  We might be able to prune it down some but a lot of these tests will remain valid if we ever get to a point of letting our users declare Workflows and feed them into SwiftUI this way (which is something that holds a lot of value for integrating into certain design patterns).

But overall I'm not sure we're actually going to use these `thenProceeds` to build out our workflows from data.

Richard-Gist commented on Sources/SwiftCurrent_SwiftUI/Views/WorkflowLauncher.swift:
```diff
@@ -77,6 +77,38 @@ public struct WorkflowLauncher<Content: View>: View {
+     - Parameter startingArgs: arguments passed to the first loaded `FlowRepresentable` in the underlying `Workflow`.
+     - Parameter workflow: workflow to be launched; must contain `FlowRepresentable`s of type `View`
+     */
+    public init<F: FlowRepresentable & View>(isLaunched: Binding<Bool>, startingArgs: AnyWorkflow.PassedArgs, workflow: Workflow<F>) where Content == AnyWorkflowItem {
```
We should rework this a little.  This is similar to the issue we just resolved.  `startingArgs` should have an iteration of `A` and an iteration of `where A == AnyWorkflow.PassedArgs`

Richard-Gist commented on Tests/SwiftCurrentTests/FlowRepresentableMetadataConsumerTests.swift:
```diff
@@ -0,0 +1,44 @@
+import SwiftCurrent
+
+class FlowRepresentableMetadataConsumerTests: XCTestCase {
+
```
```suggestion
```

Richard-Gist commented on Tests/SwiftCurrentTests/FlowRepresentableMetadataConsumerTests.swift:
```diff
@@ -0,0 +1,44 @@
+
+class FlowRepresentableMetadataConsumerTests: XCTestCase {
+
+
```
```suggestion
```

Richard-Gist commented on Tests/SwiftCurrent_SwiftUITests/SwiftCurrent_SwiftUITests.swift:
```diff
@@ -731,6 +731,258 @@ final class SwiftCurrent_SwiftUIConsumerTests: XCTestCase, App {
+                self.data = data
+            }
+        }
+//        struct FR3: View, FlowRepresentable, Inspectable {
```
Probably should remove?

Richard-Gist reviewed (approved):
Even though the Linter is failing, I am approving this PR because we are merging into a feature branch.  Those TODO failures will prevent us from accidentally merging into main with that code, but also allow us to move forward with getting data-driven working.
